File: mobile/lib/outils/telephone.dart
import 'package:flutter/services.dart';

/// Numéros de téléphone tchadiens.
///
/// **Miroir de `web/src/lib/telephone.ts`.** Les deux fichiers portent la même
/// règle : on ne peut pas partager de code entre Dart et TypeScript, alors ils
/// se citent mutuellement pour qu'une modification de l'un rappelle l'autre.
///
/// La validation côté application n'est pas une sécurité — le serveur revalide
/// tout. Elle sert à dire au parent ce qui ne va pas **avant** qu'il n'appuie
/// sur le bouton et n'attende une réponse réseau pour apprendre qu'il a oublié
/// un chiffre.
abstract final class Telephone {
  /// Premiers chiffres acceptés au Tchad, par opérateur.
  static const prefixes = ['3', '6', '8', '9'];

  /// Longueur d'un numéro national, sans l'indicatif +235.
  static const longueur = 8;

  static final _motif = RegExp('^[${prefixes.join()}]\\d{${longueur - 1}}\$');

  static String chiffresSeuls(String saisie) => saisie.replaceAll(RegExp(r'\D'), '');

  /// Extrait la partie nationale, quelle que soit la forme saisie.
  static String partieNationale(String saisie) {
    var chiffres = chiffresSeuls(saisie);

    if (chiffres.startsWith('00235')) {
      chiffres = chiffres.substring(5);
    } else if (chiffres.startsWith('235')) {
      chiffres = chiffres.substring(3);
    }

    return chiffres.replaceFirst(RegExp(r'^0+'), '');
  }

  static bool valide(String saisie) => _motif.hasMatch(partieNationale(saisie));

  /// Explique POURQUOI le numéro est refusé, ou `null` s'il est correct.
  ///
  /// « Numéro invalide » n'aide personne. Dire qu'il manque deux chiffres
  /// permet de corriger sans appeler l'école.
  static String? motifRefus(String saisie) {
    final national = partieNationale(saisie);

    if (national.isEmpty) return 'Entrez votre numéro de téléphone.';

    if (national.length < longueur) {
      final manquants = longueur - national.length;
      return 'Il manque $manquants chiffre${manquants > 1 ? "s" : ""}.';
    }

    if (national.length > longueur) {
      return 'Un numéro tchadien compte $longueur chiffres.';
    }

    if (!prefixes.contains(national[0])) {
      return 'Un numéro tchadien commence par ${prefixes.join(", ")}.';
    }

    return null;
  }

  /// Forme lisible : « 66 00 00 00 ».
  static String lisible(String saisie) {
    final national = partieNationale(saisie);
    final morceaux = <String>[];
    for (var i = 0; i < national.length; i += 2) {
      morceaux.add(national.substring(i, (i + 2).clamp(0, national.length)));
    }
    return morceaux.join(' ');
  }

  /// Forme internationale, celle que le serveur attend.
  static String international(String saisie) => '+235${partieNationale(saisie)}';
}

/// Met le numéro en forme pendant la frappe : « 66 00 00 00 ».
///
/// Les groupes de deux chiffres sont la façon dont un numéro tchadien se dit
/// et se relit. Sans eux, huit chiffres collés se vérifient mal — et un
/// parent qui ne peut pas relire son numéro ne remarque pas sa faute de
/// frappe.
class FormateurTelephone extends TextInputFormatter {
  @override
  TextEditingValue formatEditUpdate(TextEditingValue avant, TextEditingValue apres) {
    final chiffres = Telephone.chiffresSeuls(apres.text);
    final tronque = chiffres.length > Telephone.longueur
        ? chiffres.substring(0, Telephone.longueur)
        : chiffres;

    final tampon = StringBuffer();
    for (var i = 0; i < tronque.length; i++) {
      if (i > 0 && i % 2 == 0) tampon.write(' ');
      tampon.write(tronque[i]);
    }
    final texte = tampon.toString();

    // Le curseur reste à la fin : on n'autorise pas l'édition au milieu, ce
    // qui éviterait de recalculer une position après insertion d'espaces —
    // un exercice dont toutes les implémentations naïves se trompent.
    return TextEditingValue(
      text: texte,
      selection: TextSelection.collapsed(offset: texte.length),
    );
  }
}

File: web/src/lib/telephone.ts
/**
 * Numéros de téléphone tchadiens.
 *
 * Règle unique du projet, définie ICI et nulle part ailleurs. Elle est
 * dupliquée une seule fois, dans `mobile/lib/outils/telephone.dart`, faute de
 * pouvoir partager du code entre TypeScript et Dart — les deux fichiers se
 * citent mutuellement pour qu'une modification de l'un rappelle l'autre.
 *
 * Le plan de numérotation tchadien : indicatif +235, huit chiffres, dont le
 * premier identifie l'opérateur.
 */

/**
 * Premiers chiffres acceptés.
 *
 * Élargir cette liste est un changement d'un caractère ; la restreindre
 * bloquerait des parents dont le numéro est parfaitement valide. En cas de
 * doute sur un préfixe, mieux vaut l'accepter : le SMS partira, ou il ne
 * partira pas, mais le parent ne sera pas arrêté à la saisie.
 */
export const PREFIXES_TCHAD = ["3", "6", "8", "9"] as const;

/** Longueur d'un numéro national, sans l'indicatif. */
export const LONGUEUR_NATIONALE = 8;

const MOTIF = new RegExp(`^[${PREFIXES_TCHAD.join("")}]\\d{${LONGUEUR_NATIONALE - 1}}$`);

/** Ne conserve que les chiffres. */
export function chiffresSeuls(saisie: string): string {
  return saisie.replace(/\D/g, "");
}

/**
 * Extrait la partie nationale d'une saisie, quelle qu'en soit la forme.
 *
 * Le secrétariat saisit les numéros de vingt façons : « 66 00 00 00 »,
 * « +235 66000000 », « 0023566000000 », « 235-66-00-00-00 ». Toutes doivent
 * mener au même numéro, sans quoi le même parent existe en double dans le
 * fichier et ne reçoit jamais ses alertes.
 */
export function partieNationale(saisie: string): string {
  let chiffres = chiffresSeuls(saisie);

  // Préfixe international, sous ses trois écritures courantes.
  if (chiffres.startsWith("00235")) chiffres = chiffres.slice(5);
  else if (chiffres.startsWith("235")) chiffres = chiffres.slice(3);

  // Un zéro de tête est une habitude française sans effet au Tchad.
  chiffres = chiffres.replace(/^0+/, "");

  return chiffres;
}

/**
 * Vrai si la saisie correspond à un numéro tchadien complet et plausible.
 *
 * Ne vaut que pour le Tchad : l'application des parents ne propose pas de
 * sélecteur de pays, et le plan de numérotation d'un autre pays n'est pas
 * vérifiable ici. Un numéro étranger saisi au secrétariat passe par le
 * portail, pas par cet écran.
 */
export function telephoneValide(saisie: string): boolean {
  return MOTIF.test(partieNationale(saisie));
}

/**
 * Message expliquant POURQUOI le numéro est refusé.
 *
 * « Numéro invalide » n'aide personne. Dire qu'il manque deux chiffres, ou que
 * le numéro ne peut pas commencer par 5, permet de corriger sans appeler
 * l'école.
 */
export function motifRefus(saisie: string): string | null {
  const national = partieNationale(saisie);

  if (national.length === 0) return "Entrez votre numéro de téléphone.";

  if (national.length < LONGUEUR_NATIONALE) {
    const manquants = LONGUEUR_NATIONALE - national.length;
    return `Il manque ${manquants} chiffre${manquants > 1 ? "s" : ""}.`;
  }

  if (national.length > LONGUEUR_NATIONALE) {
    return `Un numéro tchadien compte ${LONGUEUR_NATIONALE} chiffres.`;
  }

  if (!PREFIXES_TCHAD.includes(national[0] as (typeof PREFIXES_TCHAD)[number])) {
    return `Un numéro tchadien commence par ${PREFIXES_TCHAD.join(", ")}.`;
  }

  return null;
}

/**
 * Numéro étranger explicitement indiqué.
 *
 * Un tuteur peut vivre hors du Tchad — un parent expatrié, un oncle en
 * France. S'il écrit son indicatif avec un « + », on le respecte : forcer
 * +235 devant produirait un numéro inexistant, et l'alerte d'absence
 * n'arriverait jamais.
 */
function etrangerExplicite(saisie: string): string | null {
  const nettoye = saisie.trim();
  if (!nettoye.startsWith("+") && !nettoye.startsWith("00")) return null;

  const chiffres = chiffresSeuls(nettoye).replace(/^00/, "");
  if (chiffres.startsWith("235")) return null;

  return chiffres.length >= 8 ? `+${chiffres}` : null;
}

/**
 * Forme internationale, seule forme stockée en base.
 *
 * Stocker « 66000000 » ici et « +235 66 00 00 00 » là rendrait impossible de
 * retrouver un parent par son numéro — et deux fiches finiraient par exister
 * pour la même personne.
 */
export function formatInternational(saisie: string): string {
  return etrangerExplicite(saisie) ?? `+235${partieNationale(saisie)}`;
}

/** Forme lisible : « 66 00 00 00 ». */
export function formatLisible(saisie: string): string {
  const national = partieNationale(saisie);
  return national.replace(/(\d{2})(?=\d)/g, "$1 ").trim();
}
